File: Antidote/NotificationView.m
//
//  NotificationView.m
//  Antidote
//
//  Created by Dmytro Vorobiov on 27.06.15.
//  Copyright (c) 2015 dvor. All rights reserved.
//

#import <Masonry/Masonry.h>

#import "NotificationView.h"
#import "UIColor+Utilities.h"
#import "AppearanceManager.h"

const CGFloat kNotificationViewHeight = 50.0;

static const CGFloat kIndentation = 10.0;
static const CGFloat kLabelsIndentation = -5.0;

@interface NotificationView ()

@property (copy, nonatomic) NotificationObject *object;

@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) UILabel *topLabel;
@property (strong, nonatomic) UILabel *bottomLabel;

@end

@implementation NotificationView

#pragma mark -  Public

- (instancetype)initWithObject:(NotificationObject *)object
{
    self = [super init];

    if (! self) {
        return nil;
    }

    self.object = object;
    self.backgroundColor = [UIColor uColorWithRed:65 green:56 blue:57 alpha:0.9];

    [self createSubviews];
    [self installConstraints];

    return self;
}

#pragma mark -  Private

- (void)createSubviews
{
    AppearanceManager *appearance = [AppContext sharedContext].appearance;

    self.imageView = [UIImageView new];
    self.imageView.backgroundColor = [UIColor clearColor];
    self.imageView.image = self.object.image;
    [self addSubview:self.imageView];

    self.topLabel = [UILabel new];
    self.topLabel.text = self.object.topText;
    self.topLabel.textColor = [UIColor whiteColor];
    self.topLabel.backgroundColor = [UIColor clearColor];
    self.topLabel.font = [appearance fontHelveticaNeueLightWithSize:16.0];
    [self addSubview:self.topLabel];

    self.bottomLabel = [UILabel new];
    self.bottomLabel.text = self.object.bottomText;
    self.bottomLabel.textColor = [UIColor whiteColor];
    self.bottomLabel.backgroundColor = [UIColor clearColor];
    self.bottomLabel.font = [appearance fontHelveticaNeueLightWithSize:16.0];
    [self addSubview:self.bottomLabel];
}

- (void)installConstraints
{
    [self.imageView makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(kIndentation);
        make.centerY.equalTo(self);
        CGFloat size = self.object.image ? kNotificationObjectImageSize : 0.0;
        make.width.height.equalTo(size);
    }];

    [self.topLabel makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.top);
        make.left.equalTo(self.imageView.right).offset(kIndentation);
        make.right.equalTo(self.right).offset(-kIndentation);
    }];

    [self.bottomLabel makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.topLabel.bottom).offset(kLabelsIndentation);
        make.left.equalTo(self.topLabel.left);
        make.right.equalTo(self.right).offset(-kIndentation);
        make.bottom.equalTo(self.bottom);
        make.height.equalTo(self.topLabel.height);
    }];
}

@end

File: Antidote/NotificationView.swift
//
//  NotificationView.swift
//  Antidote
//
//  Created by Dmytro Vorobiov on 20.01.16.
//  Copyright © 2016 dvor. All rights reserved.
//

import UIKit

class NotificationView: UIView {
    struct Constants {
        static let Offset = 10.0
        static let LabelsOffset = -5.0

        static let ImageSize: CGFloat = 34.0
        static let CloseButtonWidth = 34.0
    }

    private let tapHandler: Void -> Void
    private let closeHandler: Void -> Void

    private var imageView: UIImageView!
    private var topLabel: UILabel!
    private var bottomLabel: UILabel!

    private var fullSizeButton: UIButton!
    private var closeButton: UIButton!

    init(theme: Theme, image: UIImage, topText: String, bottomText: String, tapHandler: Void -> Void, closeHandler: Void -> Void) {
        self.tapHandler = tapHandler
        self.closeHandler = closeHandler

        super.init(frame: CGRectZero)

        backgroundColor = theme.colorForType(.NotificationBackground)

        createViews(theme: theme)
        installConstraints()

        imageView.image = image
        topLabel.text = topText
        bottomLabel.text = bottomText
    }

    required convenience init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension NotificationView {
    func fullSizeButtonPressed() {
        tapHandler()
    }

    func closeButtonPressed() {
        closeHandler()
    }
}

private extension NotificationView {
    func createViews(theme theme: Theme) {
        imageView = UIImageView()
        imageView.backgroundColor = .clearColor()
        imageView.layer.cornerRadius = Constants.ImageSize / 2
        imageView.layer.masksToBounds = true
        addSubview(imageView)

        topLabel = UILabel()
        topLabel.textColor = theme.colorForType(.NotificationText)
        topLabel.backgroundColor = .clearColor()
        topLabel.font = UIFont.systemFontOfSize(16.0)
        addSubview(topLabel)

        bottomLabel = UILabel()
        bottomLabel.textColor = theme.colorForType(.NotificationText)
        bottomLabel.backgroundColor = .clearColor()
        bottomLabel.font = UIFont.systemFontOfSize(16.0, weight: UIFontWeightLight)
        addSubview(bottomLabel)

        fullSizeButton = UIButton()
        fullSizeButton.addTarget(self, action: "fullSizeButtonPressed", forControlEvents: .TouchUpInside)
        addSubview(fullSizeButton)

        let image = UIImage.templateNamed("notification-close")

        closeButton = UIButton()
        closeButton.setImage(image, forState: .Normal)
        closeButton.tintColor = theme.colorForType(.NotificationText)
        closeButton.addTarget(self, action: "closeButtonPressed", forControlEvents: .TouchUpInside)
        addSubview(closeButton)
    }

    func installConstraints() {
        imageView.snp_makeConstraints {
            $0.left.equalTo(self).offset(Constants.Offset)
            $0.centerY.equalTo(self)
            $0.size.equalTo(Constants.ImageSize)
        }

        topLabel.snp_makeConstraints {
            $0.top.equalTo(self)
            $0.left.equalTo(imageView.snp_right).offset(Constants.Offset)
        }

        bottomLabel.snp_makeConstraints {
            $0.top.equalTo(topLabel.snp_bottom).offset(Constants.LabelsOffset)
            $0.left.equalTo(topLabel)
            $0.right.equalTo(topLabel)
            $0.bottom.equalTo(self)
        }

        fullSizeButton.snp_makeConstraints {
            $0.edges.equalTo(self)
        }

        closeButton.snp_makeConstraints {
            $0.left.equalTo(topLabel.snp_right)
            $0.right.equalTo(self)
            $0.top.bottom.equalTo(self)
            $0.width.equalTo(Constants.CloseButtonWidth)
        }
    }
}
